sway-lsp/src/capabilities/on_enter.rs

```diff
@@ -0,0 +1,207 @@
+pub(crate) async fn on_enter(
+    config: &OnEnterConfig,
+    client: &Client,
+    session: &Arc<Session>,
+    temp_uri: &Url,
+    params: &DidChangeTextDocumentParams,
+) {
+    if !(params.content_changes[0].text.contains(NEWLINE)) {
+        return;
+    }
+
+    let mut workspace_edit = None;
+    let text_document = session
+        .get_text_document(temp_uri)
+        .expect("could not get text document");
+
+    if config.continue_doc_comments.unwrap_or(false) {
+        workspace_edit = get_comment_workspace_edit(DOC_COMMENT_START, params, &text_document);
+    } else if config.continue_comments.unwrap_or(false) {
+        workspace_edit = get_comment_workspace_edit(COMMENT_START, params, &text_document);
+    }
+
+    // Apply any edits.
+    if let Some(edit) = workspace_edit {
+        if let Err(err) = client.apply_edit(edit).await {
+            tracing::error!("on_enter failed to apply edit: {}", err);
+        }
+    }
+}
+
+fn get_comment_workspace_edit(
+    start_pattern: &str,
+    change_params: &DidChangeTextDocumentParams,
+    text_document: &TextDocument,
+) -> Option<WorkspaceEdit> {
+    let range = change_params.content_changes[0]
+        .range
+        .expect("change is missing range");
+    let line = text_document.get_line(range.start.line.try_into().unwrap());
+    if line.trim().starts_with(start_pattern) {
+        let uri = change_params.text_document.uri.clone();
+        let text = change_params.content_changes[0].text.clone();
+
+        let indentation = &line[..line.find(start_pattern).unwrap()];
+        let mut edits = vec![];
+
+        // To support pasting multiple lines in a comment, we need to add the comment start pattern after each newline,
+        // except the last one.
+        let lines: Vec<_> = text.split(NEWLINE).collect();
+        lines.iter().enumerate().for_each(|(i, _)| {
+            if i < lines.len() - 1 {
+                let position = Position::new(
+                    range.start.line + (i as u32) + 1,
+                    indentation.len().try_into().unwrap(),
+                );
+                edits.push(OneOf::Left(TextEdit {
+                    new_text: format!("{start_pattern} "),
+                    range: Range::new(position, position),
+                }));
+            }
+        });
+        let edit = TextDocumentEdit {
+            text_document: OptionalVersionedTextDocumentIdentifier {
+                // Use the original uri to make updates, not the temporary one from the session.
+                uri,
+                version: None,
+            },
+            edits,
+        };
+        Some(WorkspaceEdit {
+            document_changes: Some(DocumentChanges::Edits(vec![edit])),
+            ..Default::default()
+        })
+    } else {
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use tower_lsp::lsp_types::{
+        AnnotatedTextEdit, TextDocumentContentChangeEvent, VersionedTextDocumentIdentifier,
+    };
+
```

Comment:
nit: I like use statements to not have any free lines between them.

Reply:
I agree that it's better for imports to take up less space. My strongest opinion about import formatting is that it should be done by the formatter/linter. 

In the past I've used a scheme where imports are grouped by std/extern/internal, with no lines between the imports in a group and 1 empty line between groups. Rust fmt [supports](https://rust-lang.github.io/rustfmt/?version=v1.4.38&search=#group_imports) this, but it's unstable.

I didn't find any [clippy lints](https://rust-lang.github.io/rust-clippy/master/index.html#) that do this either. I could write one...